zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1205 @@
+                    }
+                    self.nodes
+                        .try_push(Node::new(who.clone(), stake))
+                        .map_err(|_| LiquidityTreeError::TreeIsFull.into_dispatch::<T>())?;
+                    (index, BenchmarkInfo::Leaf)
+                }
+                NextNode::None => {
+                    return Err(LiquidityTreeError::TreeIsFull.into_dispatch::<T>());
+                }
+            };
+            self.account_to_index
+                .try_insert(who.clone(), index)
+                .map_err(|_| LiquidityTreeError::TreeIsFull.into_dispatch::<T>())?;
+            (index, benchmark_info)
+        };
+        if let Some(parent_index) = self.parent_index(index) {
+            self.update_descendant_stake(parent_index, stake, false)?;
+        }
+        Ok(benchmark_info)
+    }
+
+    fn exit(&mut self, who: &T::AccountId, stake: BalanceOf<T>) -> DispatchResult {
+        let index = self.map_account_to_index(who)?;
+        self.propagate_fees_to_node(index)?;
+        let node = self.get_node_mut(index)?;
+        ensure!(node.fees == Zero::zero(), LiquidityTreeError::UnclaimedFees.into_dispatch::<T>());
+        node.stake = node
+            .stake
+            .checked_sub(&stake)
+            .ok_or(LiquidityTreeError::InsufficientStake.into_dispatch::<T>())?;
+        if node.stake == Zero::zero() {
+            node.account = None;
+            self.abandoned_nodes
+                .try_push(index)
+                .map_err(|_| LiquidityTreeError::AccountNotFound.into_dispatch::<T>())?;
+            let _ = self.account_to_index.remove(who);
+        }
+        if let Some(parent_index) = self.parent_index(index) {
+            self.update_descendant_stake(parent_index, stake, true)?;
+        }
+        Ok(())
+    }
+
+    fn split(
+        &mut self,
+        _sender: &T::AccountId,
+        _receiver: &T::AccountId,
+        _amount: BalanceOf<T>,
+    ) -> DispatchResult {
+        Err(Error::<T>::NotImplemented.into())
+    }
+
+    fn deposit_fees(&mut self, amount: BalanceOf<T>) -> DispatchResult {
+        let root = self.get_node_mut(0u32)?;
+        root.lazy_fees = root.lazy_fees.checked_add_res(&amount)?;
+        Ok(())
+    }
+
+    fn withdraw_fees(&mut self, who: &T::AccountId) -> Result<BalanceOf<T>, DispatchError> {
+        let index = self.map_account_to_index(who)?;
+        self.propagate_fees_to_node(index)?;
+        let node = self.get_node_mut(index)?;
+        let fees = node.fees;
+        node.fees = Zero::zero();
+        Ok(fees)
+    }
+
+    fn shares_of(&self, who: &T::AccountId) -> Result<BalanceOf<T>, DispatchError> {
+        let index = self.map_account_to_index(who)?;
+        let node = self.get_node(index)?;
+        Ok(node.stake)
+    }
+
+    fn total_shares(&self) -> Result<BalanceOf<T>, DispatchError> {
+        let root = self.get_node(0u32)?;
+        root.total_stake()
+    }
```

Comment:
I admit that it would have been better to name everything shares. However, the naming pattern is the following:

- The tree has _stake_.
- The abstract `LiquiditySharesManager` has shares.
- The tree implements `LiquiditySharesManager` by defining shares as stake.